src/main/java/frc/robot/subsystems/SwerveSubsystem.java

```diff
@@ -0,0 +1,191 @@
+    private BooleanLogEntry controlOrientationLog = new BooleanLogEntry(datalog, "/swerve/orientation"); //Logs if robot is in FOD/ROD
+
+    public boolean controlOrientationIsFOD;
+
+    public SwerveSubsystem() {
+        new WaitUntilCommand(this::gyroReady)
+        .andThen(new InstantCommand(this::zeroHeading,this))
+        .schedule();
+        controlOrientationIsFOD = true;
+    }
+
+    private boolean gyroReady() {
+        return !gyro.isCalibrating();
+    }
+
+    public void zeroHeading() {
+        gyro.reset();
+    }
+
+    public double getHeading() {
+        //The math for this remainder is gyro - (360 * Math.round(gyro/360))
+        return Math.IEEEremainder(gyro.getAngle(), 360); //Clamps angle output between -180 and 180 degrees
+    }
+
+    public Rotation2d getRotation2d() {
+        return Rotation2d.fromDegrees(getHeading());
+    }
+
+    public Pose2d getPose() {
+        return odometer.getPoseMeters();
+    }
+
+    public void resetOdometry(Pose2d pose) {
+        odometer.resetPosition(pose, getRotation2d());
+    }
+
+    public void toggleOrientation(){
+        //Toggle control orientation from FOD/ROD
+        controlOrientationIsFOD = !controlOrientationIsFOD;
+        controlOrientationLog.append(controlOrientationIsFOD);
+    }
+
+    public SwerveModuleState[] convertToModuleStates(double xTranslation, double yTranslation, double rotation) {
+        //Takes axis input from joysticks and returns an array of swerve module states
+
+        double x = yTranslation; //Intentional, x in swerve kinematics is y on the joystick
+        double y = xTranslation;
+        double r = rotation;
+
+        //Map to speeds in meters/radians per second
+        x *= DriveConstants.kPhysicalMaxSpeed;
+        y *= DriveConstants.kPhysicalMaxSpeed;
+        r *= DriveConstants.kPhysicalMaxAngularSpeed;
+
+        //Log speeds used to construct chassis speeds
+        translationXOutputLog.append(x);
+        translationYOutputLog.append(y);
+        rotationOutputLog.append(r);
+
+        //Construct Chassis Speeds
+        ChassisSpeeds chassisSpeeds;
+        if(controlOrientationIsFOD){
+            //Field Oriented Drive
+            //TODO: test on real hardware, adding to gyro is supposed to eliminate skew of movement when rotating
+            chassisSpeeds = ChassisSpeeds.fromFieldRelativeSpeeds(x, y, r, this.getRotation2d().plus(new Rotation2d((r * Robot.kDefaultPeriod)/2)));
+        } else {
+            //Robot Oriented Drive
+            chassisSpeeds = new ChassisSpeeds(x, y, r);
+        }
+        SmartDashboard.putString("chassis speeds",chassisSpeeds.toString());
+        //Convert Chassis Speeds to individual module states
+        return DriveConstants.kDriveKinematics.toSwerveModuleStates(chassisSpeeds);  
+    }
+
+    private void normalizeDrive(SwerveModuleState[] desiredStates, ChassisSpeeds speeds){
```

Comment:
This seems like a good candidate for a unit test using JUnit to validate it does what you expect.
There is an example of this type of test in src\test\java\ExampleSubsystemTest.java

Reply:
I created a unit test on a features/swerve-junit branch as a starting point.